### DISCUSSION TITLE
Support native source folder for volume mount in remote model (#8016)

### DISCUSSION THREAD

**Discussion by fogfish** [containers/podman#13537] [General]
The discussion about an issue #8016

**Comment by aslilac** [comment]
Can we lock that issue, now that we have a discuss to talk in instead?

**Reply by baude** [reply]
done

**Comment by vanyasvl** [comment]
I have an issue with inline sed on 9p virtfs mount under MacOS with latest podman:
```
sed -i 's/aaa/zzz/g' /Users/user/q/123
sed: preserving permissions for ‘/Users/user/q/sedPtcM87’: Protocol not supported
```

Mount on podman machine:
```
vol0 on /Users/user type 9p (rw,relatime,sync,dirsync,access=client,trans=virtio)
```

part of strace:
```
read(3, "qqq\nzzz\nsss\n", 8192)        = 12
newfstatat(4, "", {st_mode=S_IFREG|0600, st_size=0, ...}, AT_EMPTY_PATH) = 0
read(3, "", 8192)                       = 0
fchown(4, 501, 1000)                    = 0
fgetxattr(3, "system.posix_acl_access", 0x7ffd6dbd18b0, 132) = -1 ENODATA (No data available)
newfstatat(3, "", {st_mode=S_IFREG|0644, st_size=12, ...}, AT_EMPTY_PATH) = 0
fsetxattr(4, "system.posix_acl_access", "\2\0\0\0\1\0\6\0\377\377\377\377\4\0\4\0\377\377\377\377 \0\4\0\377\377\377\377", 28, 0) = -1 EPROTONOSUPPORT (Protocol not supported)
fsetxattr(4, "system.posix_acl_access", "\2\0\0\0\1\0\6\0\377\377\377\377\4\0\4\0\377\377\377\377 \0\4\0\377\377\377\377", 28, 0) = -1 EPROTONOSUPPORT (Protocol not supported)
```
Issue is in `fsetxattr` syscall

**Reply by fogfish** [reply]
Thank you for reporting it! I can reproduce it as well. It seems to be well know issue at QEMU https://lists.gnu.org/archive/html/qemu-devel/2020-12/msg02655.html

Let's re-open an issue, otherwise it would not be resolved.

**Comment by shanesmith** [comment]
What is the expected ownership of bind mounted volumes in a container?

I would expect it would that that of the container's `USER` (or `--user` flag in `podman run`), but I see that the macOS user is kept instead in a rootful container.

In the example below, the file is owned by shane (501) on macOS and that is reflected in the Podman container, unlike Docker which sets the owner to root.

```
$ ls -la adir
total 0
drwxr-xr-x   3 shane  staff   96  6 Apr 13:21 .
drwxr-xr-x  16 shane  staff  512  6 Apr 13:14 ..
-rw-r--r--   1 shane  staff    0  6 Apr 13:20 afile

$ podman machine init -v $HOME:$HOME --now

$ podman -c podman-machine-default-root run -v $PWD/adir:/adir alpine ls -la /adir
total 0
drwxr-xr-x    3 501      dialout         96 Apr  6  2022 .
dr-xr-xr-x    1 root     root            40 Apr  5 12:02 ..
-rw-r--r--    1 501      dialout          0 Apr  6  2022 afile

$ docker run -v $PWD/adir:/adir alpine ls -la /adir
total 4
drwxr-xr-x    3 root     root            96 Apr  6 17:21 .
drwxr-xr-x    1 root     root          4096 Apr  6 17:25 ..
-rw-r--r--    1 root     root             0 Apr  6 17:20 afile

```

**Reply by shanesmith** [reply]
@baude sorry to be pinging you directly, but I wonder if you have any thoughts on the above? =)

**Comment by baude** [comment]
@shanesmith well docker has a significant advantage here in that they can just assume root for everything.  we do not have the luxury.  we def fiddle with permissions to make them usable by the core fcos user int he vm, and like you observed, we try to set the uid (number) to the same as the macos uid, or you run into neverending permissions errors.  Perhaps it might be best if you were to say what is wrong with our approach and we see about making you happy?

**Reply by shanesmith** [reply]
I don't have enough filesystem experience to argue what's wrong, mostly using Docker as a starting point because that where we (and I'm sure lots of people) are coming from.

I've played around with Docker Desktop some more and found some surprising behaviour. Even though the files inside the mounted directory are owned by root (or `-u USER`), all users in a containers are able to perform any operation inside this directory. In other words, file ownership and permissions are completely ignored. For example, reading/writing/deleting files and folders owned by root and set to user-only permissions (ie: `chmod 600`) by a normal user is allowed. Also surprising is the fact that some operations that only touch/read files inside the mounted directory will actually change the owner of those files, although I guess this makes no practical difference since again any user can still do any operation.

This setup seems to make sense for Docker Desktop since it's only used for development and removing potential headaches like file permissions is desirable, especially considering the user could be less technical either generally or in container technologies. What's odd though is that I can't find any mention of this behaviour online, either in Docker's documentation or otherwise...

In any case, now the question is would it make sense for Podman to adopt this behaviour whenever Podman Machine is used? =)

**Reply by rhatdan** [reply]
Definitely not.  We want to maintain security if possible.  I wonder if this lack of permission checks is because Docker is using a primative version of virtiofsd for volume mounting.  We are working with the upstream qemu and virtiofsd teams to get proper support in.
I believe Docker is using something provided by the Apple virtulization stack based on virtiofs.

**Reply by shanesmith** [reply]
Docker Desktop currently uses grpc-fuse but does have experimental virtiofs support. I've tried both and found the same behaviour. It does feel like it's something they do on purpose.

Just spitballing, would it be possible to enable this behaviour in Podman through a flag of some sort?

**Reply by rhatdan** [reply]
No because we are relying on the underlying file system.

**Reply by Sangeppato** [reply]
@baude Sorry to directly ping you on this.
The current behavior causes some issues with VSCode's dev-containers, since when re-opening a mac-local folder in a Podman container using a non-root user, the folder will actually be impossible to write to.
I personally think that Podman's behavior makes more sense in general, but on a desktop where containers are only used for local development I consider Docker's behavior to be more desirable since, as @shanesmith mentioned, it doesn't require the user to fiddle with file permission on the host machine or inside the VM.

Is there any way to allow an arbitrary non-root user inside a Podman container to write to the host filesystem mounted inside the VM?

**Reply by danielhodder** [reply]
While I applaud the general stance of more secure, this issue also makes poadman incompatible with Docker when used in the desktop sense. It would be great if there was, at least, some kind of option a user could specify in the desktop app, or VM, to make it compatible with Docker. A lot of developer facing containers have been built with the assumption of Docker Desktop's broken file system permissions.

Podman is never going to convince people to change the containers, so really should have an option to ignore file permissions - if that's what the user wants. Podman desktop, specifically, is all about easing use and being a replacement for Docker Desktop (which is more important now that ever with Docker's latest price hikes).

**Reply by rhatdan** [reply]
What exactly would you want changed?  Do you want the volumes owned by root for rootfull containers or by the user for rootless containers.
Why is the volume impossible to write too, if the user within the VM is running as your UID and the files owned on the system are the same UID?

**Reply by ssbarnea** [reply]
@rhatdan IMHO, what a macos user would want would be `security_model=none` and implicit mounted volumes. Still few minutes ago I found another problem while running a container:
```
Error: error preparing container 0a3bcd84182bfece73f001a5afcb25bada1a4997cf21d26c4dba7c3866fb5ad2 for attach:
lsetxattr /var/folders/32/1xrphgzd4xv777syxjtkpdw80000gn/T/ansible-navigator_06i7pwty/artifacts/21241976-fc14-48f4-b90f-bf7892837368: operation not supported
```
As `/var/folders/` is mounted, I am not surprised that `lsetxattr` as I doubt macos apfs would support extended attributes. Now I need to figure-out where this comes from, but I am inclined to believe that the podman machine might have selinux enabled on it, something that would clearly open another class of incompatibilities.

My WIP guide about using podman on maco keeps growing while I need to add more hacks to make these work. Check https://github.com/ansible/vscode-ansible/wiki/macos

And apparently, I was right about selinux:

```
$ podman machine ssh
Connecting to vm podman-machine-default. To close connection, use `~.` or `exit`
Fedora CoreOS 36.20221001.2.0
Tracker: https://github.com/coreos/fedora-coreos-tracker
Discuss: https://discussion.fedoraproject.org/tag/coreos

$ sestatus
SELinux status:                 enabled
SELinuxfs mount:                /sys/fs/selinux
SELinux root directory:         /etc/selinux
Loaded policy name:             targeted
Current mode:                   enforcing
Mode from config file:          enforcing
Policy MLS status:              enabled
Policy deny_unknown status:     allowed
Memory protection checking:     actual (secure)
Max kernel policy version:      33
```

How can we make selinux not to mess with the mounted folders?

**Reply by rhatdan** [reply]
SELinux is enabled, but I would need to see the AVCs from within the container to know what is causing the failure.  Doing :Z or :z on a volume would fail, since it will not be allowed to relabel the volumes mounted from the Mac.

**Reply by rhatdan** [reply]
Can you try security_model=none and see if this fixes the issue.

**Reply by ssbarnea** [reply]
I mentioned that I already had to set `security_model=none` on all mounted volumes, mainly because otherwise the symlinks on them would not work.

I need to mount 3 locations from macos inside podman machine:
* `$HOME` as that is where the source code is (containers are primarily used to lint code)
* `/private/tmp` and `/var/folders/` because the tool that runs the containers also performs some mounts using subdirectories from these temps location. At least I was lucky that these locations do not match anything existing on Linux, so they would not hide location already used by linux as that would likely break something.

What I think it is happening is that selinux does inherit some of the special restrictions of `/var` and applies them to the mounted volume, causing the error that we seen with `lsetxattr`.

Any hints on what should I try?

As options, I see either fully disable selinux inside podman machine or reconfigure it to no try any selinux shenanigans on the mounted volumes.

Obviously that in longer term we will want to make these default behaviors for mac, so users will be able to start using podman on mac for development without any extra hacking - A "just works" experience, especially as on this platform 99% of the people would use containers for development and not for production, so the security requirements are quite different than on a linux system.

**Reply by rhatdan** [reply]
I am not sure it is SELinux, that is why I asked for the audit information from within the machine.

podman machine ssh ausearch -m avc

**Reply by rhatdan** [reply]
There could be just applications that are attempting to set xattrs.
Containerized applications should not be attempting to set xattrs and podman should only attempt if a user told it do using a :Z or :z on a volume mount.

**Reply by ssbarnea** [reply]
Apparently `sudo setenforce 0` is not enough as I still seen the error.

I am starting to believe that this might be a bug in podman where it does attempt to use `lsetxattr` on mounts that do not support it. But at the same time, I did not see the same type of error when I mounted $HOME.

**Reply by rhatdan** [reply]
setenforce 0, would not change the behaviour.  If Podman thinks it needs to do SELinux actions of setting labels, it will do it in enforcing mode and permissive mode, and still fail if it is not allowed.  I just need to see where it is reporting failure.

**Reply by ssbarnea** [reply]
You are right, it was not selinux. It was https://github.com/containers/podman-compose/issues/509 -- which for some reason seems closed, while I do not see it as resolved.

The presence of `:Z` would produce this. Once I removed the `:Z`, it started to work just fine.

**Reply by ssbarnea** [reply]
I proposed a workaround for ansible-runner at https://github.com/ansible/ansible-runner/pull/1146 but you can easily see how this would not scale well. What about other tools that might try to do the same.

I am wondering if we could neuter the `:Z` on macos, so we would accept the label without doing anything about it. Arguably we could print a debugging message saying that we are ignoring the label as not being supported. WDYT?

**Reply by rhatdan** [reply]
I think the best could do is drop it to logrus.Warnf when ENOSUP.

**Comment by gaba-xyz** [comment]
I'm seeing very poor I/O performance on the mounted volume inside of containers. I've posted benchmarks from my MBP i7 below. Is this expected?

## host

```console
$ dd if=/dev/zero of=$HOME/Documents/zeroes_host.bak count=1M
1048576+0 records in
1048576+0 records out
536870912 bytes (512.0MB) copied, 5.220928 seconds, 103 MB/s
```

## container

```console
$ dd if=/dev/zero of=/app/zeroes_container.bak count=1M
1048576+0 records in
1048576+0 records out
536870912 bytes (512.0MB) copied, 280.546129 seconds, 1.8MB/s
```

**Comment by samskiter** [comment]
Is there a solution to this issue identified yet? I'm seeing it on macos. is it a qemu issue? do i need to change my volume mount setup?

**Reply by bernhardriegler** [reply]
In the issue it is discussed that volumes need to be setup on init of the podman machine.

see https://github.com/containers/podman/issues/8016

also https://medium.com/@butkovic/favoring-podman-over-docker-desktop-33368e031ba0

**Reply by samskiter** [reply]
Thanks so much for the reply, i missed that! Is there a resolution for this bug coming?

**Reply by samskiter** [reply]
(not that I was seeing the mounting of volumes *partially* working but for some files was getting either permission denied or something like this error from above: `preserving permissions for ‘/Users/user/q/sedPtcM87’: Protocol not supported`

**Reply by samskiter** [reply]
in my case it happens during a `mv` command. The file is moved but the permissions presumably aren't correctly set.

**Reply by vanyasvl** [reply]
`sed -i` also doesn't work due to `Protocol not supported` error

**Comment by samskiter** [comment]
It seems this is related back to qemu and 9pfs https://gitlab.com/qemu-project/qemu/-/issues/944

**Comment by arduino-man** [comment]
I am being affected by this issue as well. Was very happy to move to Podman from Docker but this issues might just be what makes me go back :(

**Comment by samariation** [comment]
same issue also with me :(

**Comment by o-az** [comment]
I didn't even get to spend a full day with Podman now I'm having to go back to Docker. SMH

**Comment by ssbarnea** [comment]
Here is something to think about, please refrain from making comments that do not add any value to the discussion, they only spam and alienate others at best, making less likely to want to engage in finding solutions for the identified problems. If you just like/agree or disagree/dislike something, use the buttons, not comments.

We could try to help others by hiding some useless comments but is draining. Yes, I am frustrated from time to time about the experience I get while using podman but I also know that podman team was one of the most open to suggestions and welcoming pull requests, very different than the other solution we are still competing with. Remember that podman is as free as it gets, you don't need a license or subscription to use it. But your time is too valuable for debugging these issue, pay someone to help you but don't troll others.

Here is my current guide on podman tricks on macos: https://github.com/ansible/vscode-ansible/wiki/macos - remoting is not covered yet because I know is not working, also I know that is not really working even with docker. If you need mounting from local, using remoting is not suitable, performance wise anyway.. to easy to end-up wanting to mount gigabytes of data, even by mistake.

**Reply by pykenny** [reply]
@ssbarnea I saw the comment [here](https://stackoverflow.com/a/74013039/8589324), which says:

> Is likely that most MacOS users would only want to mount **from within their home directory,** ...

Does the solution work on external drive, or any volume outside of home directory? Working on a MacBook with limited storage left, so ideally I want both the virtual machine and containers to use external drive for storage.

**Reply by pykenny** [reply]
Just saw the notes in the tutorial:

> You will not be able to mount directories outside your MacOS home directory

So does that mean running Podman on MacOS with external drive is still (entirely?) not doable?

**Comment by rhatdan** [comment]
This is by default, you can add additional directories via the podman machine init command.
```
man podman-machine-init
...
   --volume, -v=source:target[:options]
       Mounts a volume from source to target.

       Create a mount. If /host-dir:/machine-dir is specified as the *source:target*, Podman mounts host-
       dir in the host to machine-dir in the Podman machine.

       Additional options may be specified as a comma-separated string. Recognized  options  are:  *  ro:
       mount  volume  read-only * rw: mount volume read/write (default) * security_model=[model]: specify
       9p security model (see below)

       The   9p   security    model    [determines]    https://wiki.qemu.org/Documentation/9psetup#Start‐
       ing_the_Guest_directly  if  and how the 9p filesystem translates some filesystem operations before
       actual storage on the host.

       In order to allow symlinks to work, on MacOS the default security model is
        none.

       The value of mapped-xattr specifies that 9p store symlinks and some file  attributes  as  extended
       attributes  on  the host. This is suitable when the host and the guest do not need to interoperate
       on the shared filesystem, but has caveats for actual shared access; notably, symlinks on the  host
       are  not usable on the guest and vice versa. If interoperability is required, then choose none in‐
       stead, but keep in mind that the guest will not be able to do things that  the  user  running  the
       virtual machine cannot do, e.g. create files owned by another user. Using none is almost certainly
       the best choice for read-only volumes.

       Example: -v "$HOME/git:$HOME/git:ro,security_model=none"

       Default volume mounts are defined in containers.conf.   Unless  changed,  the  default  values  is
       $HOME:$HOME.

```

**Reply by pykenny** [reply]
Tried mapping my external drive to VM's system root:

```shell
podman machine init -v="/Volumes/my_extern_drive_root:/"
```
➡️ 
```text
Extracting compressed file
Image resized.
Machine init complete
To start your machine run:

	podman machine start
```
---
```shell
podman machine start
```
➡️ 
```text
Starting machine "podman-machine-default"
Waiting for VM ...
Mounting volume... /Volumes/my_extern_drive_root:/
WARN[0039] API socket failed ping test

This machine is currently configured in rootless mode. If your containers
require root permissions (e.g. ports < 1024), or if you run into compatibility
issues with non-podman clients, you can switch using the following command:

	podman machine set --rootful

API forwarding listening on: /Users/pykenny/.local/share/containers/podman/machine/podman-machine-default/podman.sock

The system helper service is not installed; the default Docker API socket
address can't be used by podman. If you would like to install it run the
following commands:

	sudo /usr/local/Cellar/podman/4.3.1/bin/podman-mac-helper install
	podman machine stop; podman machine start
```
---
```shell
sudo /usr/local/Cellar/podman/4.3.1/bin/podman-mac-helper install

podman machine stop; podman machine start
```
➡️ 
```text
Waiting for VM to exit...
Machine "podman-machine-default" stopped successfully
Starting machine "podman-machine-default"
Waiting for VM ...
Mounting volume... /Volumes/my_extern_drive_root:/
Error: exit status 255
```

🤷 

 - The VM is still in running status, even though getting error message from `machine-start` command (which means doing `podman machine stop` right after still works).
 - Rebuild the VM (`podman machine rm; podman machine init; podman machine start`) seems to avoid the error instead of restart existing VM.
   But then you got this error on `podman machine ssh`:
   ```text
   Connecting to vm podman-machine-default. To close connection, use `~.` or `exit`
   Fedora CoreOS 37.20230122.2.0
   Tracker: https://github.com/coreos/fedora-coreos-tracker
   Discuss: https://discussion.fedoraproject.org/tag/coreos

   cannot chdir to /var/home/core: No such file or directory
   ```

**Reply by pykenny** [reply]
Seems that `--volume` option only handles volume mapping, but doesn't deal with location to install and run the VM.

**Comment by pykenny** [comment]
Okay, so let's forget about the plan to move ~~everything~~ anything (including the VM) to remote drive, and try following the instructions in "Volumn" chapter of *Podman in Action*, with few modifications to fit Podman's remote mode (Podman 4.3.1, MacOS 13.2.1):

1️⃣ Map VM's `$HOME/sandbox/podman` directory to mines in the host system:
```shell
podman machine init -v "$HOME/sandbox/podman:$HOME/sandbox/podman"
```
(Reference: `man podman-machine-init`:)
```text
   --volume, -v=source:target[:options]
       Mounts a volume from source to target.


       Create a mount. If /host-dir:/machine-dir is specified as the *source:target*, Podman
       mounts host-dir in the host to machine-dir in the Podman machine.
```

➡️ 
```text
Extracting compressed file
Image resized.
Machine init complete
To start your machine run:

	podman machine start
```

---
2️⃣ Start VM:
Run `podman machine start` ➡️ 🆗 

---
3️⃣ Run example image and map `/var/www/html` directory in the container, to VM's `$HOME/sandbox/podman/html`, which is supposed to be `$HOME/sandbox/podman/html` in my host system beneath.
```shell
podman run -d -v "$HOME/sandbox/podman/html:/var/www/html:ro,z" -p 8080:8080 quay.io/rhatdan/myimage
```

(Reference: `man podman-run`:)
```text
   --volume, -v=[[SOURCE-VOLUME|HOST-DIR:]CONTAINER-DIR[:OPTIONS]]
       Create a bind mount. If -v /HOST-DIR:/CONTAINER-DIR is specified, Podman bind mounts
       /HOST-DIR from the host into /CONTAINER-DIR in the Podman container. Similarly, -v
       SOURCE-VOLUME:/CONTAINER-DIR will mount the named volume from the host into the
       container. If no such named volume exists, Podman will create one. If no source is
       given, the volume will be created as an anonymously named volume with a randomly
       generated name, and will be removed when the container is removed via the --rm flag
       or the podman rm --volumes command.


       (Note when using the remote client, including Mac and Windows (excluding WSL2)
       machines, the volumes will be mounted from the remote server, not necessarily the
       client machine.)
```

➡️ ⚡⚡⚡ 
```
Trying to pull quay.io/rhatdan/myimage:latest...
Getting image source signatures
Copying blob sha256:e3460238f8a1f4698e1ec867ff96682f5d45debdd10e0503742fd15124d8bf5b
Copying blob sha256:c7765172d3ce59f229d53f5c2a60346ad3922c29baa7ae19d31ef9866117d743
Copying blob sha256:dfd8c625d0226c52da48ce402e79bc6e60a360d732bb7f6523c62cb714ec0a0d
Copying blob sha256:2b782a9ad894d15e65ee92d0e294b8358cfc69d94bfd5b2cf8d5d286376a0f4a
Copying blob sha256:a1eadb69adf1f7b62f76fc7bc2d7f8c28e6c03dc1f6024a4f9fd1329412efc89
Copying config sha256:2c7e43d880382561ebae3fa06c7a1442d0da2912786d09ea9baaef87f73c29ae
Writing manifest to image destination
Storing signatures
Error: lsetxattr /(MY_$HOME)/sandbox/podman/html: operation not supported
```

**Reply by pykenny** [reply]
Reason for forgetting about all the stuff about using external drive:

Just saw one another message from @ssbarnea in the [same document](https://github.com/ansible/vscode-ansible/wiki/macos) above:

> You will **not** be able to mount directories **outside your MacOS home directory**, so don't try to open workspaces hosted on other disks or network drivers, as podman will not be able to mount them.

😭

**Reply by pykenny** [reply]
By default, `podman-machine-start` maps host system's `$HOME` to VM's `$HOME` unless we overwrite the configuration. However and unfortunately, the same error will still happen for the example above.

**Reply by pykenny** [reply]
Also, [changing mount type](https://github.com/ansible/vscode-ansible/wiki/macos#3-correct-podman-machine-mount-types) before starting VM does not work as well with the same error message emitted

 - Update: Also tried adding `/private/tmp:/private/tmp` and `/var/folders/` in mount settings. Still out of luck.

**Reply by pykenny** [reply]
Tried upgrading Podman to 4.4.1 (`brew upgrade podman`), `podman-machine-rm` the VM, and finally redo all steps. Doesn't help.

**Reply by pykenny** [reply]
Finally, removing `z` flag (which MacOS doesn't support) in mount option works:
 - https://github.com/containers/podman/issues/13631

**Comment by ssbarnea** [comment]
Based on my tests, installing 4.4.1 removing old podman machine and reinstalling it using default settings did add the expected 3 folders to the machine mount, the two macos temp ones and `/Users`, which should be fine for most users.

**Reply by pykenny** [reply]
Not my case, unfortunately. But seems that `podman-run`'s documentation has changed ~~a lot~~ compared to its 4.3 companion. So maybe the configuration syntax has changed and I'm reading `podman-run`'s manpage again now. (By the way, _Podman in Action_ was written during the 4.1 era)

**Reply by pykenny** [reply]
<img width="600" alt="image" src="https://user-images.githubusercontent.com/6328813/219531957-cf96450f-ab78-4f72-9301-c253ecd931a1.png"/>

Also the new manpage's broken on my console (compared to [rendered HTML](https://docs.podman.io/en/latest/markdown/podman-run.1.html#security-opt-option))

**Reply by pykenny** [reply]
[Solved](https://github.com/containers/podman/discussions/13537#discussioncomment-5003597). Documentation should mention about that.

**Comment by awidjaja** [comment]
In my case, podman machine is initialized using podman desktop on Windows, so this workaround don't work. I use a separate WSL distribution for my local development environment. By adding podman connection setup, I can run podman remotely using that podman machine. But how can I mount my local development directory (not the host home dir) into the podman machine? I think the mounting should be done upon connection add, rather than machine init. If possible, the podman client should open a tunnel which expose the local directory contents for the podman machine to access remotely.

**Reply by rdbisme** [reply]
I'm in the same situation. Did we raise a ticket?

**Reply by jankap** [reply]
Similar discussion as here: https://github.com/containers/podman/discussions/13537

But I'm facing the same issue, the problem is that I can't access WSL mounts to be volume mounted when doing podman init. 

Are there any ways to do this on runtime, i.e. mount volumes after init?

**Comment by ecerulm** [comment]
I was experiencing this problem today using podman 4.5.1 on macOS Ventura 13.4. 

Restarting the podman machine should generally work , I can see that the default podman machine already mounts `/Users`, `/private` and `/var/folders`

```
podman machine stop
podman machine start 
Starting machine "podman-machine-default"
Waiting for VM ...
Mounting volume... /Users:/Users
Mounting volume... /private:/private
Mounting volume... /var/folders:/var/folders
podman run -ti --rm -v "$HOME:/myhome" ubi8 
[root@ba2054e82401 /]# ls /myhome
```

If the podman machine it's not mounting those you can recreate and try again, but that deletes all your containers, images, etc. 

```
podman machine stop
podman machine rm
podman machine init # optionally put init options here like --memory 8192
podman machine start
Starting machine "podman-machine-default"
Waiting for VM ...
Mounting volume... /Users:/Users
Mounting volume... /private:/private
Mounting volume... /var/folders:/var/folders
podman run -ti --rm -v "$HOME:/myhome" ubi8 
[root@ba2054e82401 /]# ls /myhome
```

**Comment by rdbisme** [comment]
Hello, this makes  `devcontainers` impossible to use when developing from inside another WSL VM. The volume mount that is passed to `podman-remote` is looked for on the server instead of the client, therefore the repo is not possible to be exposed correctly. 

I guess this makes it very hard to run `podman-remote` commands with `-v` binds: it requires statically defined mounts in the `podman-machine` VM. 

Is there a way to expose filesystem access between host<->remote at runtime? I think I've read about using `SSHFS` or maybe something similar to GRPC-fuse which I think docker uses?

**Reply by rdbisme** [reply]
I keep stumbling in myself of the past which keeps suggesting the same things :D. 

This seems to be the active issue: https://github.com/containers/podman/issues/21813
